Add repeats to Timer.collect_callgrind(...)

robieta opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53296 Definition infrastructure for instruction count ubenchmarks
* **#53295 Add repeats to Timer.collect_callgrind(...)**

A lot of the time spent in `collect_callgrind` is spinning up Valgrind and executing the initial `import torch`. In most cases the actual run loop is a much smaller fraction. As a result, we can reuse the same process to do multiple replicates and do a much better job amortizing that startup cost. This also tends to result in more stable measurements: the kth run is more repeatable than the first because everything has been given a chance to settle into a steady state. The instruction microbenchmarks lean heavily on this behavior. I found that in practice doing several `n=100` replicates to be more reliable than one monolithic 10,000+ iteration run. (Since rare cases like memory consolidation will just contaminate that one replicate, as opposed to getting mixed into the entire long run.)

Differential Revision: [D26907093](https://our.internmc.facebook.com/intern/diff/D26907093)

Commit:
Add repeats to Timer.collect_callgrind(...)

[ghstack-poisoned]

Commit:
Update on "Add repeats to Timer.collect_callgrind(...)"




[ghstack-poisoned]

Commit:
Update on "Add repeats to Timer.collect_callgrind(...)"




[ghstack-poisoned]

Commit:
Update on "Add repeats to Timer.collect_callgrind(...)"




[ghstack-poisoned]

Commit:
Update on "Add repeats to Timer.collect_callgrind(...)"


A lot of the time spent in `collect_callgrind` is spinning up Valgrind and executing the initial `import torch`. In most cases the actual run loop is a much smaller fraction. As a result, we can reuse the same process to do multiple replicates and do a much better job amortizing that startup cost. This also tends to result in more stable measurements: the kth run is more repeatable than the first because everything has been given a chance to settle into a steady state. The instruction microbenchmarks lean heavily on this behavior. I found that in practice doing several `n=100` replicates to be more reliable than one monolithic 10,000+ iteration run. (Since rare cases like memory consolidation will just contaminate that one replicate, as opposed to getting mixed into the entire long run.)

[ghstack-poisoned]

Commit:
Update on "Add repeats to Timer.collect_callgrind(...)"


A lot of the time spent in `collect_callgrind` is spinning up Valgrind and executing the initial `import torch`. In most cases the actual run loop is a much smaller fraction. As a result, we can reuse the same process to do multiple replicates and do a much better job amortizing that startup cost. This also tends to result in more stable measurements: the kth run is more repeatable than the first because everything has been given a chance to settle into a steady state. The instruction microbenchmarks lean heavily on this behavior. I found that in practice doing several `n=100` replicates to be more reliable than one monolithic 10,000+ iteration run. (Since rare cases like memory consolidation will just contaminate that one replicate, as opposed to getting mixed into the entire long run.)

[ghstack-poisoned]

ngimel reviewed (approved):
That's a very nice improvement, both repeat and streamlining of baseline collection!

Commit:
Update on "Add repeats to Timer.collect_callgrind(...)"


A lot of the time spent in `collect_callgrind` is spinning up Valgrind and executing the initial `import torch`. In most cases the actual run loop is a much smaller fraction. As a result, we can reuse the same process to do multiple replicates and do a much better job amortizing that startup cost. This also tends to result in more stable measurements: the kth run is more repeatable than the first because everything has been given a chance to settle into a steady state. The instruction microbenchmarks lean heavily on this behavior. I found that in practice doing several `n=100` replicates to be more reliable than one monolithic 10,000+ iteration run. (Since rare cases like memory consolidation will just contaminate that one replicate, as opposed to getting mixed into the entire long run.)

Differential Revision: [D26907093](https://our.internmc.facebook.com/intern/diff/D26907093)

[ghstack-poisoned]

pbelevich commented:
Reverted because:
```
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 581, in wrapper
    fn(*args, **kwargs)
  File "benchmark_utils/test_benchmark_utils.py", line 519, in test_collect_callgrind
    self.assertGreater(max(counts.values), 1, f"Every instruction count total was unique: {counts}")
TypeError: 'builtin_function_or_method' object is not iterable
```

robieta commented:
Gah. Sorry about that. (And thanks for treehugging.)
